### PR TITLE
Add molecular_dipole function 

### PR DESCRIPTION
**Pull request by ddhawan11**
**Context:**
Add molecular_dipole function to calculate dipole moment operator in Pauli basis.

**Description of the Change:**
molecular_dipole function is analogous to molecular_hamiltonian function to calculate dipole moment operator with openfermion and dhf backends

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by ddhawan11** [comment]
[sc-62858]

File: pennylane/qchem/openfermion_obs.py [hunk 1]
```diff
@@ -161,6 +161,15 @@ def observable(fermion_ops, init_term=0, mapping="jordan_wigner", wires=None):
             openfermion.transforms.bravyi_kitaev(mb_obs), wires=wires
         )
 
+    if mapping == "parity":
+        qubits = openfermion.count_qubits(mb_obs)
+        if qubits == 0:
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:166]: When do we expect this to happen? 🤔

Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py:166]: Sometimes one of the x,y,z components of the dipole moment operator are 0, when mapped this should return I(0). For openfermion, when such an operator is passed to count_qubits, it gives 0.

File: pennylane/qchem/openfermion_obs.py [hunk 2]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+            For type dict, only int-keyed dict (for qubit-to-wire conversion) is accepted for
+            partial mapping. If None, will use identity map.
+        args (array[array[float]]): initial values of the differentiable parameters
+        cutoff (float): Cutoff value for including the matrix elements
+            :math:`\langle \alpha \vert \hat{{\bf r}} \vert \beta \rangle`. The matrix elements
+            with absolute value less than ``cutoff`` are neglected.
+
+    Returns:
+        list[pennylane.Hamiltonian]: the qubit observables corresponding to the components
+        :math:`\hat{D}_x`, :math:`\hat{D}_y` and :math:`\hat{D}_z` of the dipole operator.
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:739]: We never mention anything about dipole being a vector quantity in the description above. We should do that.

File: pennylane/qchem/openfermion_obs.py [hunk 3]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+    if mapping.strip().lower() not in ["jordan_wigner", "parity", "bravyi_kitaev"]:
+        raise ValueError(
+            f"'{mapping}' is not supported."
+            f"Please set the mapping to 'jordan_wigner', 'parity' or 'bravyi_kitaev'."
+        )
+
+    symbols = molecule.symbols
+    coordinates = molecule.coordinates
+
+    if len(coordinates) == len(symbols) * 3:
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:756]:
```suggestion
    if qml.math.shape(coordinates)[0] == len(symbols) * 3:
```

File: pennylane/qchem/openfermion_obs.py [hunk 4]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+        geometry_dhf = qml.numpy.array(coordinates.reshape(len(symbols), 3))
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:757]: Is there any particular reason why we are using `qml.numpy` and not `qml.math`?

File: pennylane/qchem/openfermion_obs.py [hunk 5]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+        geometry_hf = coordinates
+    elif len(coordinates) == len(symbols):
+        geometry_dhf = qml.numpy.array(coordinates)
+        geometry_hf = coordinates.flatten()
+
+    if molecule.mult != 1:
+        raise ValueError(
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:764]: Is this true also for `openfermion` backend? 🤔

File: pennylane/qchem/openfermion_obs.py [hunk 6]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+        cutoff (float): Cutoff value for including the matrix elements
+            :math:`\langle \alpha \vert \hat{{\bf r}} \vert \beta \rangle`. The matrix elements
+            with absolute value less than ``cutoff`` are neglected.
+
+    Returns:
+        list[pennylane.Hamiltonian]: the qubit observables corresponding to the components
+        :math:`\hat{D}_x`, :math:`\hat{D}_y` and :math:`\hat{D}_z` of the dipole operator.
+
+
+    """
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:742]: We should add an example to the docs.

File: pennylane/qchem/openfermion_obs.py [hunk 7]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+
+    core, active = qml.qchem.active_space(
+        molecule.n_electrons, molecule.n_orbitals, molecule.mult, active_electrons, active_orbitals
+    )
+
+    if method == "dhf":
+
+        if args is None and isinstance(geometry_dhf, qml.numpy.tensor):
+            geometry_dhf.requires_grad = False
+        mol = qml.qchem.Molecule(
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:782]: On a second thought, do we need to make the whole `Moelcule` again or should we add the processing regarding the coordinates to the molecule class?

File: pennylane/qchem/openfermion_obs.py [hunk 8]
```diff
@@ -649,6 +658,169 @@ def dipole_of(
...
+            load_data=molecule.load_data,
+            alpha=molecule.alpha,
+            coeff=molecule.coeff,
+        )
+
+        requires_grad = args is not None
+        dip = (
+            qml.qchem.dipole_moment(mol, cutoff=cutoff, core=core, active=active, mapping=mapping)(
+                *args
+            )
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:797]: We could get rid of `if` statement by doing this -
```suggestion
        dip = (
            qml.qchem.dipole_moment(mol, cutoff=cutoff, core=core, active=active, mapping=mapping)(
                *(args or [])
            )
```

File: tests/qchem/of_tests/test_molecular_dipole.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        method="dhf",
+        active_electrons=active_el,
+        active_orbitals=active_orb,
+        mapping=mapping,
+        outpath=tmpdir.strpath,
+    )
+
+    for idx, dip in enumerate(dip_dhf):
+        wires = dip.wires
+        if not wires:
```
Comment (obliviateandsurrender) [inline tests/qchem/of_tests/test_molecular_dipole.py:297]: Is this when we have an empty dipole op?

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Thanks @ddhawan11, leaving the first round of comments!

File: pennylane/qchem/openfermion_obs.py (hunk 5 above)
Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py:758]: Yes, this one doesn't work for openshell systems even for openfermion backend.

File: pennylane/qchem/openfermion_obs.py (hunk 7 above)
Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py:776]: This changes the basis_set and r as well, depending on whether gradients are required for coordinates or not.

File: tests/qchem/of_tests/test_molecular_dipole.py (hunk 9 above)
Comment (ddhawan11) [reply tests/qchem/of_tests/test_molecular_dipole.py:297]: yes

File: pennylane/qchem/openfermion_obs.py (hunk 4 above)
Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py:751]: not really, I just kept it the way it was, is there a benefit to using qml.math?

Comment (soranjh) [reply pennylane/qchem/openfermion_obs.py:781]: Functions in qml.math are framework-agnostic, i.e., they can detect the type of the array/tensor and use proper framework. We use qml.math when possible.

File: pennylane/qchem/openfermion_obs.py [hunk 10]
```diff
@@ -802,6 +1001,7 @@ def decompose(hf_file, mapping="jordan_wigner", core=None, active=None):
 def molecular_hamiltonian(*args, **kwargs):
     """molecular_hamiltonian(molecule, method="dhf", active_electrons=None, active_orbitals=None,\
     mapping="jordan_wigner", outpath=".", wires=None, args=None, convert_tol=1e12)
+
```
Comment (soranjh) [inline pennylane/qchem/openfermion_obs.py:1004]: Why this is changes? The PR is not synced with master?

File: pennylane/qchem/openfermion_obs.py [hunk 11]
```diff
@@ -857,6 +1057,7 @@ def molecular_hamiltonian(*args, **kwargs):
         The ``molecular_hamiltonian`` function accepts a ``Molecule`` object as its first argument.
         Look at the `Usage Details` for more details on the old interface.
 
+
```
Comment (soranjh) [inline pennylane/qchem/openfermion_obs.py:1060]: Why this is changes? The PR is not synced with master?

File: tests/qchem/of_tests/test_molecular_dipole.py [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+ops_h2.append([I(0)])
+ops_h2.append([I(0)])
+ops_h2.append(
+    [
+        Y(0) @ Z(1) @ Y(2),
+        X(0) @ Z(1) @ X(2),
+        Y(1) @ Z(2) @ Y(3),
+        X(1) @ Z(2) @ X(3),
+    ]
+)
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_dipole.py:43]: Please mention this, if true:
# the reference dipole operator

File: tests/qchem/of_tests/test_molecular_dipole.py [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+
+eig_h2 = []
+eig_h2.append([0.0, 0.0])
+eig_h2.append([0.0, 0.0])
+eig_h2.append([-1.81780062, -0.90890031, -0.90890031])
+
+
+h3p = ["H", "H", "H"]
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_dipole.py:69]: Are these the same tests as in other original functions?

File: tests/qchem/of_tests/test_molecular_dipole.py [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+ops_h2_parity.append([I(0)])
+ops_h2_parity.append([I(0)])
+ops_h2_parity.append(
+    [
+        Y(0) @ Y(1),
+        X(0) @ X(1) @ Z(2),
+        Y(1) @ Y(2),
+        Z(0) @ X(1) @ X(2) @ Z(3),
+    ]
+)
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_dipole.py:61]: Where these data come from?

**Review by soranjh — COMMENTED** [review]
Thanks @ddhawan11, this looks great. Please address my minor comments/questions and I will approve.

File: tests/qchem/of_tests/test_molecular_dipole.py (hunk 13 above)
Comment (ddhawan11) [reply tests/qchem/of_tests/test_molecular_dipole.py:69]: These tests are mostly same as from the original functions, I have added a couple of more tests, one for wires mapping, also for parity mapping in both cases.

File: tests/qchem/of_tests/test_molecular_dipole.py (hunk 14 above)
Comment (ddhawan11) [reply tests/qchem/of_tests/test_molecular_dipole.py:61]: This data is from the old tests.

File: pennylane/qchem/openfermion_obs.py [hunk 15]
```diff
@@ -649,6 +658,196 @@ def dipole_of(
...
+    elif len(coordinates) == len(symbols):
+        geometry_dhf = qml.numpy.array(coordinates)
+        geometry_hf = coordinates.flatten()
+
+    if molecule.mult != 1:
+        raise ValueError(
+            "Open-shell systems are not supported. Change the charge or spin multiplicity of the molecule."
+        )
+
+    wires_map = None
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:792]: Why do we define this above? 🤔 More importantly, `wires_map` is only being used for `qml.map_wires`, so why not use it directly there?

Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py]: Good point, thanks for noticing this.

File: pennylane/qchem/openfermion_obs.py (hunk 10 above)
Comment (ddhawan11) [reply pennylane/qchem/openfermion_obs.py]: This seems to be a sync issue as I started from qchem_feature branch, I have changed this.

**Review by soranjh — APPROVED** [review]
Thanks @ddhawan11, looks good to go.

File: pennylane/qchem/openfermion_obs.py [hunk 16]
```diff
@@ -649,6 +658,193 @@ def dipole_of(
...
+        dip = (
+            qml.qchem.dipole_moment(mol, cutoff=cutoff, core=core, active=active, mapping=mapping)(
+                *(args or [])
+            )
+            if requires_grad
+            else qml.qchem.dipole_moment(
+                mol, cutoff=cutoff, core=core, active=active, mapping=mapping
+            )()
+        )
+        wires_map = None
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:821]:
```suggestion
```
Why do we need to define this beforehand if we don't use it except within the `if` context?

**Review by obliviateandsurrender — APPROVED** [review]
Thanks @ddhawan11! Good to go once the final comment is resolved!

File: doc/code/qml_qchem.rst
Comment (soranjh) [inline doc/code/qml_qchem.rst:1]: Why so many changes?

**Review by soranjh — CHANGES_REQUESTED** [review]

File: doc/code/qml_qchem.rst
Comment (ddhawan11) [reply doc/code/qml_qchem.rst:1]: This is as I am in between resolving conflicts.

File: doc/code/qml_qchem.rst [hunk 17]
```diff
@@ -194,4 +194,4 @@ the user with
 .. code-block:: bash
 
     pip install openfermionpyscf
-    pip install pyscf
\ No newline at end of file
+    pip install pyscf
```
Comment (soranjh) [inline doc/code/qml_qchem.rst:197]: Anything changed here?

**Review by soranjh — APPROVED** [review]

File: doc/code/qml_qchem.rst (hunk 17 above)
Comment (ddhawan11) [reply doc/code/qml_qchem.rst]: Nothing has changed here.
